### PR TITLE
Added EraGen instance for Alonzo

### PR DESCRIPTION
**Pull request by TimSheard**
Aded an EraGen instance for Alonzo in the Tests:   Test.Cardano.Ledger.Alonzo.EraGenInstance

I had to generalize the EraGen class slightly to make an Alonzo instance adding the method genEraPParamsDelta
because the type family PParamsDelta  is different in Alonzo than in earlier Eras.
I had to export a few functions in a few files where they were no exported before, because they were needed in the Alonzo instance.  Made few other small changes to a number of files.

### CODE REVIEW

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs [hunk 1]
```diff
@@ -575,3 +578,32 @@ getLanguageView ::
...
+
+-- | Turn an PParams' into a Shelley.Params'
+retractPP :: (HKD f Coin) -> PParams' f era2 -> Shelley.PParams' f era1
+retractPP
+  c
+  (PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau d eE pv mnP _ _ _ _ _ _) =
+    (Shelley.PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau d eE pv c mnP)
+
+-- | Given the missing pieces Turn a Shelley.PParams' into an Params'
+extendPP ::
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:632]: This already exists as `translatePParams` in `Cardano.Ledger.Alonzo.Translation`

File: shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs [hunk 2]
```diff
@@ -93,6 +93,8 @@ type instance Core.PParams (ShelleyEra c) = SPP.PParams (ShelleyEra c)
 
 type instance Core.Witnesses (ShelleyEra c) = WitnessSet (ShelleyEra c)
 
+type instance Core.PParamsDelta (ShelleyEra c) = SPP.PParamsUpdate (ShelleyEra c)
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs]: This is already defined on line 75 above. The fact that it works being defined there is a ltitle scary, since the type family is no longer associated with the class, but if adding it in another place we should remove it from the line above.

Comment (nc6) [reply shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs]: Oh, even worse! There are *two* `PParamsDelta`, since https://github.com/input-output-hk/cardano-ledger-specs/commit/c497f64ed73fd4c8783aa63835065c63e7634568 didn't delete the old one when the new one was added! We should definitely clean this up, otherwise it's very confusing (and potentially harmful)

**Review by nc6 — CHANGES_REQUESTED** [review]

File: shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs (hunk 2 above)
Comment (TimSheard) [reply shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs]: cleaned up.

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs (hunk 1 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PParams.hs:632]: extendPP works for either Id or StrictMaybe, so it is more general. So I kept it, and rewrote both translatePParams and  translatePParamsUpdate as a one-liners:

translatePParams (AlonzoGenesis ada cost price mxTx mxBl mxV) pp = 
 extendPP pp ada cost price mxTx mxBl mxV

translatePParamsUpdate pp = extendPP pp SNothing SNothing SNothing SNothing SNothing SNothing

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Trace/Chain.hs [hunk 3]
```diff
@@ -62,39 +62,41 @@ import Test.Shelley.Spec.Ledger.ConcreteCryptoTypes
...
-  ( EraGen (..),
-    genUtxo0,
-  )
+import Test.Shelley.Spec.Ledger.Generator.Core (GenEnv (..))
+
 import Test.Shelley.Spec.Ledger.Generator.Presets (genesisDelegs0)
-import Test.Shelley.Spec.Ledger.Generator.Update (genPParams)
-import Test.Shelley.Spec.Ledger.Shrinkers (shrinkBlock)
+-- import Test.Shelley.Spec.Ledger.Shrinkers (shrinkBlock) -- TODO FIX ME
 import Test.Shelley.Spec.Ledger.Utils
```
Comment (TimSheard) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Trace/Chain.hs:69]: Eventually we will need to add some sort of Shrinkers back.

**Comment by TimSheard** [comment]
This PR touches a lot of files, but the vast majority are only changing the types of functions that were over constrained to Shelley-like Eras.

It adds instances so the Alonzo Era can now run the property tests.

It added a bunch of missing PrettyA instances.

File: alonzo/impl/src/Cardano/Ledger/Alonzo.hs [hunk 4]
```diff
@@ -214,6 +199,7 @@ instance CC.Crypto c => EraModule.SupportsSegWit (AlonzoEra c) where
   fromTxSeq = Alonzo.txSeqTxns
   toTxSeq = Alonzo.TxSeq
   hashTxSeq = Alonzo.hashTxSeq
+  unsafeApplyTx (Tx bod wit auxdata) = ValidatedTx bod wit (IsValidating True) auxdata
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo.hs]: This method should not exist in this class

File: cardano-ledger-core/src/Cardano/Ledger/Era.hs [hunk 5]
```diff
@@ -133,6 +138,10 @@ class SupportsSegWit era where
     TxSeq era ->
     Hash.Hash (CryptoClass.HASH (Crypto era)) EraIndependentBlockBody
 
+  -- | Use unsafeApplyTx only for Tests. The real applyTx is an STS rule
+  -- that performs phase 1 validation, as well as the injection into TxInBlock.
+  unsafeApplyTx :: Core.Tx era -> TxInBlock era
```
Comment (nc6) [inline cardano-ledger-core/src/Cardano/Ledger/Era.hs]: No, this should not live here. If it absolutely is needed, there should be a separate class that exists only in the test libraries

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs [hunk 6]
```diff
@@ -153,7 +153,15 @@ collectTwoPhaseScriptInputs pp tx utxo =
     Just cost -> merge (apply cost) (map redeemer needed) (map getscript needed) (Right [])
   where
     txinfo = transTx utxo tx
-    needed = scriptsNeeded utxo tx
+    needed = filter knownToNotBe1Phase $ scriptsNeeded utxo tx
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs]: What does this have to do with testing?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs [hunk 7]
```diff
@@ -158,6 +158,8 @@ instance
   (Era era, Core.AnnotatedData (Core.Script era)) =>
   Semigroup (WitnessSetHKD Identity era)
   where
+  (WitnessSet' a b c _) <> y | Set.null a && Map.null b && Set.null c = y
+  y <> (WitnessSet' a b c _) | Set.null a && Map.null b && Set.null c = y
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:162]: The point of these is just to avoid unnecessary serialisation?

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Delegation.hs [hunk 8]
```diff
@@ -89,6 +86,11 @@ import Test.Shelley.Spec.Ledger.Generator.Core
   )
 import Test.Shelley.Spec.Ledger.Generator.EraGen (EraGen (..))
 import Test.Shelley.Spec.Ledger.Utils
+import GHC.Records(HasField(..))
+import Shelley.Spec.Ledger.BaseTypes(UnitInterval)
+import Shelley.Spec.Ledger.PParams(ProtVer)
```
Comment (nc6) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Delegation.hs:91]: Ordering

**Review by nc6 — CHANGES_REQUESTED** [review]

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs (hunk 6 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs]: this is a bug fix (that I have posted in https://github.com/input-output-hk/cardano-ledger-specs/pull/2272 as well). Tim was trying to generate valid Alonzo Txs, and I told him that that might help.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs (hunk 7 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:162]: The point was to be able to have a `Semigroup` instance that would not dangerously remove the memo-bytes if you multiplied by the identity.

**Review by TimSheard — COMMENTED** [review]
this is a mistake

File: alonzo/impl/src/Cardano/Ledger/Alonzo.hs (hunk 4 above)
Comment (TimSheard) [reply alonzo/impl/src/Cardano/Ledger/Alonzo.hs]: Moved to another class. Good idea.

File: cardano-ledger-core/src/Cardano/Ledger/Era.hs (hunk 5 above)
Comment (TimSheard) [reply cardano-ledger-core/src/Cardano/Ledger/Era.hs]: done

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Delegation.hs (hunk 8 above)
Comment (TimSheard) [reply shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Delegation.hs:91]: Ormolise will fix this.

**Review by JaredCorduan — APPROVED** [review]
Thank you @TimSheard for taking on this huge and unpleasant task!

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 9]
```diff
@@ -185,7 +185,7 @@ data UtxoPredicateFailure era
   | TriesToForgeADA
   | -- | list of supplied bad transaction outputs
     OutputTooBigUTxO
-      ![Core.TxOut era]
+      ![(Int, Int, Core.TxOut era)]
```
Comment (nc6) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:188]: Should label in haddock what these parameters mean

**Review by nc6 — APPROVED** [review]
